example-async.js

```diff
@@ -0,0 +1,155 @@
+'use strict'
+
+const Fastify = require('fastify')
+
+function build (opts) {
+  const fastify = Fastify(opts)
+
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb-async' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply) {
+    const jwt = this.jwt
+    const level = this.level
+
+    if (request.body && request.body.failureWithReply) {
+      reply.code(401).send({ error: 'Unauthorized' })
+      return Promise.reject(new Error())
+    }
+
+    if (!request.req.headers['auth']) {
+      return Promise.reject(new Error('Missing token header'))
+    }
+
+    return new Promise(function (resolve, reject) {
+      jwt.verify(request.req.headers['auth'], function (err, decoded) {
+        if (err) { reject(err) };
+        resolve(decoded)
+      })
+    }).then(function (decoded) {
+      return level.get(decoded.user)
+        .then(function (password) {
+          if (!password || password !== decoded.password) {
+            throw new Error('Token not valid')
+          }
+        })
+        .catch(function (err) {
+          if (err.notFound) {
+            throw new Error('Token not valid')
+          }
+          throw err
```

Comment:
This err is going to be caught by the external catch,  and it will never be surfaced. That’s ok, but that means we can remove this catch completely.

Reply:
got you, thanks, done